### PR TITLE
Add build progress to console plain mode

### PR DESCRIPTION
**Pull request by kmark**
This pull request adds occasional progress indicators to the `plain` console mode.

### Context
Right now, it’s impossible to get the build progress when using the plain console mode. This is useful to have when `tail -f`’ing build logs from a CI pipeline, where you need to use plain mode but want the human friendly progress indications.

Implementation info/decisions are included in the commit messages. Unopinionated on the actual format of the progress updates.

Thanks!

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

gradle-plain-progress

### CODE REVIEW

**Comment by jbartok** [comment]
Thank you for giving this a try. Since is such a visual change, would it be possible for you to provide a screen-cast illustrating the before-after state of your changes? 

***

Maybe use https://asciinema.org/ to do it?

**Comment by kmark** [comment]
Rebased and pushed up a fix to make it less noisy (as originally intended).

> would it be possible for you to provide a screen-cast illustrating the before-after state of your changes?

Sure, here’s a screencast of running `classes testClasses` (realistic workload) on this repo.

[asciicast](https://asciinema.org/a/vr608FDCpGWuA0WMoY9yR3eGK)

We could make it a little less chatty by adding a minimum monotonic time between progress updates. But for any reasonably sized build the progress updates even in the worst case will be a small fraction of the total log lines.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
> Hey @jbartok, let me know if there’s anything else you need from me on this one. Thanks!

Seems like the change is messing with a bunch of tests that rely on plain console output... I'll get back to you on how we should approach this.

**Comment by kmark** [comment]
Thanks @jbartok. Happy to update the tests. I haven't done so yet to ensure we were comfortable with the feature and approach before moving on to fixing all of the tests.

**Comment by jbartok** [comment]
> Thanks @jbartok. Happy to update the tests. I haven't done so yet to ensure we were comfortable with the feature and approach before moving on to fixing all of the tests.

Ok, please go ahead then and update the tests.

**Comment by kmark** [comment]
@jbartok On it. Do you have any opinions on if this feature should be default off or default on? That has some implications for how I’ll go about fixing up the tests, beyond just the user facing changes.

**Comment by jbartok** [comment]
> Do you have any opinions on if this feature should be default off or default on?

I would say default OFF.

**Comment by kmark** [comment]
Added this feature as an opt-in behavior via `--console=plainProgress`. Still need to add integration tests but is otherwise usable.

**Comment by kmark** [comment]
Tests are added. `:sanityCheck` and `:logging:quickTest` are passing. Ready for another look whenever you get the chance @jbartok.

File: subprojects/docs/src/docs/userguide/running-builds/command_line_interface.adoc [hunk 1]
```diff
@@ -607,11 +607,13 @@ You can control the use of rich output (colors and font variants) by specifying
 `-Dorg.gradle.console=(auto,plain,rich,verbose)`::
 Specify console mode via Gradle properties. Different modes are described immediately below.
 
-`--console=(auto,plain,rich,verbose)`::
+`--console=(auto,plain,plainProgress,rich,verbose)`::
 Specifies which type of console output to generate.
 +
 Set to `plain` to generate plain text only. This option disables all color and other rich output in the console output. This is the default when Gradle is _not_ attached to a terminal.
 +
+Set to `plainProgress` to generate plain text only like `plain` with occasional progress updates.
```
Comment (jbartok) [inline subprojects/docs/src/docs/userguide/running-builds/command_line_interface.adoc:615]:
```suggestion
Set to `plainProgress` to generate plain text like `plain`, but with extra (occasional) progress updates.
```

File: subprojects/logging-api/src/main/java/org/gradle/api/logging/configuration/ConsoleOutput.java [hunk 2]
```diff
@@ -24,6 +28,13 @@ public enum ConsoleOutput {
      * Disable all color and rich output. Generate plain text only.
      */
     Plain,
+    /**
+     * Disable all color and rich output. Generate plain text only. Occasionally reports progress.
+     *
+     * @since 8.4
```
Comment (jbartok) [inline subprojects/logging-api/src/main/java/org/gradle/api/logging/configuration/ConsoleOutput.java:34]: This won't make it into Gradle `8.4`; please change this to `8.5` (and others, if there are any).

**Comment by kmark** [comment]
Rebased on latest master and applied requested edits.

**Comment by jbartok** [comment]
Hey @kmark.

Sorry for the delayed response. We had some internal discussion on this again; there were some concerns. Ultimately, we agreed to make this functionality part of the `plain` console and NOT add an extra console mode.

I know you already put a lot of effort into it, but would you be willing to push further along that line?

**Review by lkasso — APPROVED** [review]

**Review by lkasso — COMMENTED** [review]
Ignore approval.

**Comment by kmark** [comment]
Hey @jbartok,

No worries. I appreciate the team's time.

I had this originally tied into the plain console mode so rolling the recent changes back shouldn't be much trouble.

Do we still want to provide a command line option to toggle this behavior within the plain console mode or are we okay with this becoming a permanent always on feature?

**Comment by jbartok** [comment]
> Hey @jbartok,
> 
> No worries. I appreciate the team's time.
> 
> I had this originally tied into the plain console mode so rolling the recent changes back shouldn't be much trouble.
> 
> Do we still want to provide a command line option to toggle this behavior within the plain console mode or are we okay with this becoming a permanent always on feature?

Please make it a permanent, always-on feature.

**Comment by kmark** [comment]
Hey @jbartok,

Rebased against current `master` and re-integrated the feature back into the normal `progress` console mode. Let me know what you all think.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle cancel

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
Hi @kmark. Would you mind looking into the failures above? You should be able to access our Team City instance as a guest, let me know if not.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by kmark** [comment]
Hey @jbartok, fixed up output scraping and the Windows-specific test.

**Comment by jbartok** [comment]
Thanks @kmark. Let me try to push this through CI today.

**Comment by kmark** [comment]
Looks like 2 more to go. I can take a look at these later today.

**Comment by jbartok** [comment]
@kmark I don't understand that weird, intermittent failure in `AutoConsoleBasicGroupedTaskLoggingFunctionalTest.long running task output correctly interleave with other tasks in parallel console attached to stderr only`, will need to investigate further, when I get to it.

**Comment by kmark** [comment]
@jbartok It looks like the transient failures were due to an incomplete regex that didn’t handle the `(FAILING)` case of the progress reports.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Comment by kmark** [comment]
Fixed the `ToolingApiLoggingCrossVersionSpec` issue by just filtering out the progress logs. Not certain if that’s what we want to do, but confirmed it works.

**Comment by jbartok** [comment]
@bot-gradle test RFN

**Comment by jbartok** [comment]
@bot-gradle test RFN

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by kmark** [comment]
Instead of playing this perpetual game of whack-a-mole with these integration tests, I added progress normalization directly to the `AbstractIntegrationSpec`. That feels like a better approach.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Comment by kmark** [comment]
Appreciate the heads up @blindpirate. Based on the thread dump it looks like this is happening when we hit:

https://github.com/gradle/gradle/blob/bf2aa0b44037ddaea1e66c216d44566924745ca0/platforms/core-runtime/logging/src/main/java/org/gradle/internal/logging/console/PlainBuildStatusRenderer.java#L89-L91

and it’s deadlocking because the `Logger` is attempting to contribute to the build output pipeline.

Sounds like we need to apply the fix in #24593 here.

**Comment by jbartok** [comment]
> > Sounds like we need to apply the fix in #24593 here.
> 
> Hmm... shouldn't this branch already include the code from #24593?

Yes it is included.

**Comment by kmark** [comment]
Sorry, I should’ve been more clear. I need to apply the fix from that PR to the code I wrote here: https://github.com/gradle/gradle/blob/bf2aa0b44037ddaea1e66c216d44566924745ca0/platforms/core-runtime/logging/src/main/java/org/gradle/internal/logging/console/PlainBuildStatusRenderer.java#L89-L91

**Comment by kmark** [comment]
Added the, admittedly strange, fix for the deadlock. Overflowed progress will now show as 99% instead of 101%, 102%, etc. My deadlock fix differs slightly from the one in #24593 in that I’m taking a copy of the progress state before launching the new thread, which avoids a race condition where we might log different values than the ones we tested against.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test RFN

**Comment by kmark** [comment]
Hey @jbartok, thank you for merge, guidance, and support for this change over the past year! If you hit issues with this down the line, feel free to ping/mention me here on GitHub.

**Pull request merged**